### PR TITLE
C++: Improve handling of re-use expressions

### PR DESCRIPTION
**Pull request by jketema**
This will slightly brake IR generation for databases created with CodeQL 2.17.1. There are two changes:
* Some type are different in the IR. This seems inconsequential.
* The IR becomes inconsistent when a null pointer is being deleted. This should hopefully not occur at all in real world code.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/exprs/Expr.qll [hunk 1]
```diff
@@ -6,6 +6,7 @@ import semmle.code.cpp.Element
 private import semmle.code.cpp.Enclosing
 private import semmle.code.cpp.internal.ResolveClass
 private import semmle.code.cpp.internal.AddressConstantExpression
+private import semmle.code.cpp.internal.ExtractorVersion
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/exprs/Expr.qll:9]: This isn't needed anymore, right?

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/exprs/Expr.qll:9]: Oops.

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/exprs/Expr.qll]: Fixed

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Comment by jketema** [comment]
@MathiasVP note I added some additional text to the top-level comment.

**Comment by MathiasVP** [comment]
> @MathiasVP note I added some additional text to the top-level comment.

Thanks for adding that piece of context. I think that's fine. AFAIK, we don't have any queries that:
- Use the IR, and
- Look for null pointer dereferences like that.

So, as you said yourself, I don't think this will have any impact on real-world DBs.

**Pull request merged**